drivers: adc: Add a driver support for TI MSPM0 ADC module

Alex9360 opened the pull request:
Add driver support for TI MSPM0 ADC module with multi channel sampling support

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

Alex9360 committed:
dts: arm: ti: mspm0: Add ADC node for TI MSPM0

Add dts support for ADC node.

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

Alex9360 commented on dts/bindings/adc/ti,mspm0-adc12.yaml:
```diff
@@ -0,0 +1,85 @@
+  vref:
+    type: phandle
+
+  max_result:
```
It represents the maximum no of result storage registers. It varies for g and l series

lsd-he360 replied:
I think it is the same for g and l series but different for c series?

In any case, I think it's better to define the max in device-specific files in the ti hal and not have to require this for users?

lsd-he360 commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,534 @@
+		}
+
+		if (!(data->vref_flags & INT_VREF)) {
+			if (regulator_set_voltage(config->vref_config, vref, vref) < 0) {
```
I am not sure why I would need to compile the regulator driver if I am just using the VDD as the reference voltage. This would unnecessarily take up some amount of space on memory-constrained systems.

lsd-he360 commented on dts/bindings/adc/ti,mspm0-adc12.yaml:
```diff
@@ -0,0 +1,85 @@
+  vref:
+    type: phandle
+
+  max_result_reg:
```
@Alex9360 I can't seem to mark the comment as resolved, but the changes look good.

parthitce replied:
`max-result-reg` and also may be with `ti,` prefix?

lsd-he360 commented on dts/arm/ti/mspm0/g/mspm0gx51x.dtsi:
This is the wrong address for adc0 for MSPM0G3519.

lsd-he360 commented on drivers/adc/adc_mspm0_adc12.c:
Do you need this or can you just use `CONFIG_MSPM0_PERIPH_STARTUP_DELAY`?

lsd-he360 commented:
Please consider using `FIELD_GET` and `FIELD_PREP` macros for setting configuration settings in `channel_mem_ctl`.

lsd-he360 commented on drivers/adc/adc_mspm0_adc12.c:
`ADC12_MEMCTL_CHANSEL_OFS` is 0. What's the purpose of the bit shift?

Alex9360 commented on drivers/adc/adc_mspm0_adc12.c:
@lsd-he360  As it have only capable for setting the sampling time via the scomp register we can maximum use two different sampling time only. either we can take scomp0 or scomp1 value for each channel.

Aman-Lachhiramka-ti commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,586 @@
+	uint32_t vref = ((uint32_t)api->ref_internal) * ADC_MSPM0_VREF_MULTIPLIER;
+#endif
+
+	if (ch > ADC_MSPM0_CHANNEL_MAX) {
```
It should be  
ch >= ADC_MSPM0_CHANNEL_MAX

Alex9360 committed:
include: zephyr: dt-bindings: clock: add the hfclk macro

Add hfclk macro for adc hfclk configuration purpose

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

Alex9360 committed:
drivers: clock_control: Add hfclk struct

Add hfclk struct to get the clock rate of hfclk

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

jsbatch commented on dts/bindings/adc/ti,mspm0-adc12.yaml:
```diff
@@ -0,0 +1,81 @@
+      - 32
+      - 48
+
+  auto_powerdown:
```
```suggestion
  auto-powerdown:
```
> Use dashes (-) as word separators for node and property names.

https://docs.zephyrproject.org/latest/contribute/style/devicetree.html

jsbatch commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,586 @@
+			return wakeup_cycles;
+		}
+
+		acq_time = acq_time + wakeup_cycles;
```
The acq_time variable comes from channel_cfg->acquisition_time, which has this encoding (from adc.h):
- Bits [15:14] - unit (1=microseconds, 2=nanoseconds, 3=ticks)
- Bits [13:0] - value

So this line is adding wakeup_cycles to the encoded value, potentially corrupting the result.

jsbatch commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,586 @@
+		}
+	} else {
+		period_ns = ADC_MSPM0_NS_PER_SEC / (clock_rate / config->divider);
+		/* Wakup time is 5us for both l and g mspm0 series */
```
```suggestion
		/* Wakeup time is 5us for both l and g mspm0 series */
```

JFMSP reviewed (commented):
Thanks for getting this up! 

I have some small notes about the repeat mode enabled and some potential code clean up that could result and a few questions. 

Let me know if I can help in any way

JFMSP commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,587 @@
+		for (mem_ix = 0; mem_ix < data->channel_eoc; mem_ix++) {
+			*data->buffer++ = DL_ADC12_getMemResult(config->base, mem_ix);
+		}
+		*data->buffer = DL_ADC12_getMemResult(config->base, mem_ix);
```
Ok I see that update_buffer_pointer is called based on if we have CONTINUE or REPEAT and that will increment. If FINISH is called presumably the buffer will reset before it is called again.

JFMSP commented on drivers/adc/adc_mspm0_adc12.c:
I think auto is also appropriate for most use cases here. The hardcoding of initSeqSample with SAMPLING_SOURCE_AUTO will allow for 99% of use cases, as sampling time may still be set with the acquisition time.

JFMSP reviewed (approved):
Thanks for getting this up! 

I have some small notes about the repeat mode enabled and some potential code clean up that could result and a few questions. 

Let me know if I can help in any way

ssekar15 commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,583 @@
+	static DEVICE_API(adc, mspm0_driver_api##index) = {                                        \
+		.channel_setup = adc_mspm0_channel_setup,                                          \
+		.read = adc_mspm0_read,                                                            \
+		.ref_internal = DT_INST_PROP(index, vref_mv),                                      \
```
does the `vref_mv` is variable for each instance of ADC? if not keep `DEVICE_API` section as single instance.

otherwise `ref_internal` should not be part of `DEVICE_API`.

ssekar15 commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,583 @@
+			      &adc_mspm0_cfg_##index, POST_KERNEL, CONFIG_ADC_INIT_PRIORITY,       \
+			      &mspm0_driver_api##index);                                           \
+                                                                                                   \
+	static void adc_mspm0_cfg_func_##index(void)                                               \
```
move api def above DEVICE_DT_INST_DEFINE

ssekar15 commented on drivers/adc/Kconfig.mspm0:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 Texas Instruments
```
2026

Alex9360 committed:
tests: drivers: adc: adc_api: boards: Add overlay for test suite

Add a device tree overlay required for running the ADC API
test suite on lp_mspm0g3507 board.

Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

lsd-he360 commented on drivers/adc/adc_mspm0_adc12.c:
```diff
@@ -0,0 +1,593 @@
+		wakeup_cycles = ADC_MSPM0_WAKEUP_TIME_NS / period_ns;
+
+		if (acq_time == ADC_ACQ_TIME_DEFAULT) {
+			return wakeup_cycles;
```
If I am not mistaken, this would configure the SCOMP with only the cycles needed for wakeup but doesn't add the cycles needed to actually sample? Am I correct?

@Alex9360 @JFMSP

Alex9360 commented:
@lsd-he360 There is no explicit mention about minimum sampling clock cycles needed in auto power down mode in trm or data sheet and it is fixed by returning wakeup_cycles + 1. As per TRM  "SCOMPx should be > (5uS + Duration for sample window)". The previous code returned only wakeup_cycles which violated this by leaving zero cycles for sampling. The +1 gives 1 cycle sampling window, consistent with the non-auto_pwdn default where SCOMP=0 gives 1 cycle sample pulse per TRM  "The default SCOMPx sample timer value translates to 1 cycle wide sample pulse". Users requiring longer sampling  should provide explicit acquisition-time ticks via DTS.

what is your opinion about this @JFMSP ?

Alex9360 commented:
The dependency PR #94732 has merged successfully.

Alex9360 commented:
ping for review @ssekar15 @parthitce . Thanks!

Alex9360 commented:
@jsbatch ping for review.

parthitce reviewed (changes requested):
> dts: bindings: adc: Add support for TI MSPM0 ADC module
> Add a binding support for TI MSPM0 ADC module.

rebase is broken. dts commit included the driver itself.

Alex9360 committed:
dts: bindings: adc: Add support for TI MSPM0 ADC module

Add a binding support for TI MSPM0 ADC module.

Signed-off-by: Jackson Farley <j-farley@ti.com>
Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

Alex9360 committed:
drivers: adc: Add a driver support for TI MSPM0 ADC module

Add driver support for TI MSPM0 ADC module.

Add a configurable reference voltage support with vref.
Add this driver in shell.

Signed-off-by: Jackson Farley <j-farley@ti.com>
Signed-off-by: Alexpandi Muniyandi <alexpandi@linumiz.com>

parthitce commented on include/zephyr/dt-bindings/clock/mspm0_clock.h:
```diff
@@ -23,5 +23,7 @@
 #define MSPM0_CLOCK_CANCLK		MSPM0_CLOCK(0x7, 0x0)
 #define MSPM0_CLOCK_CLK_OUT		MSPM0_CLOCK(0x8, 0x0)
+/** @brief High frequency clock reference */
+#define MSPM0_CLOCK_HFCLK		MSPM0_CLOCK(0x9, 0x2)
```
nit, although this can be moved to the dt bindings patch, I see there is no usage for the macro , but only an option for consumer.

parthitce reviewed (approved):
Minor nit, but that's also fine. LGTM.
